FeedStoreChallenge/Infrastructure/CoreData/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,46 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				if let cache = try ManagedCache.fetch(from: context) {
+					completion(.found(
+						feed: cache.localFeed,
+						timestamp: cache.timestamp)
+					)
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let managedCache = try ManagedCache.createNewUniqueCacheInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				try ManagedCache.fetch(from: context).map(context.delete(_:)).map(context.save)
```

Comment:
Suggestion:
`(_:)` can be omitted to make the code shorter
```suggestion
				try ManagedCache.fetch(from: context).map(context.delete).map(context.save)
```

Reply:
Done